base/src/main/java/io/spine/code/proto/enrichment/EnrichmentType.java

```diff
@@ -0,0 +1,201 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.code.proto.enrichment;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.protobuf.Descriptors.Descriptor;
+import io.spine.code.proto.MessageType;
+import io.spine.code.proto.PackageName;
+import io.spine.code.proto.ref.DirectTypeRef;
+import io.spine.code.proto.ref.TypeRef;
+import io.spine.type.KnownTypes;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+
+/**
+ * An enrichment type is a message which is added to a context of another message
+ * to extends its information for the benefit of message handlers.
+ *
+ * <p>An enrichment message declaration has:
+ * <ol>
+ *   <li>The {@link io.spine.option.OptionsProto#enrichmentFor
+ *       (enrichment_for)} option with a {@linkplain TypeRef reference} to one or more message types
+ *       this message enriches.
+ *   <li>One or more fields with the {@link io.spine.option.OptionsProto#by (by)} option with
+ *       the value {@linkplain FieldRef referencing} one ore more fields of source message types.
+ * </ol>
+ *
+ * <p>If one of the above is missing, the enrichment declaration is not valid.
+ */
+public final class EnrichmentType extends MessageType {
+
+    private final ImmutableList<TypeRef> sourceTypeRefs;
+    private final ImmutableSet<MessageType> knownSources;
+    private final ImmutableList<FieldDef> fields;
+    private final ImmutableMap<MessageType, FieldMatch> fieldMatches;
+
+    @VisibleForTesting
+    static EnrichmentType from(Descriptor descriptor) {
+        EnrichmentType result = new EnrichmentType(descriptor);
+        return result;
+    }
+
+    /**
+     * Obtains an enrichment type for the passed name.
+     */
+    public static EnrichmentType from(MessageType type) {
+        checkNotNull(type);
+        Descriptor descriptor = type.descriptor();
+        return from(descriptor);
+    }
+
+    /**
+     * Verifies if the passed message type is an enrichment type.
```

Comment:
The "valid / invalid" terms aren't used here. However, there is "the enrichment declaration is not valid" above. Is this method somehow related to "valid"? Another thing is that "valid / invalid" is also about message and constraints. 

I would not use "valid" term in this class if it's not related to message constraints.

Reply:
I've updated docs. But I'd rather tell more about validity of declaration because we do validate that the declaration is supported by known types and their fields.

Reply:
Not in terms of validation framework constraints, but it's still validation. I understand that the term is too broad, though.